# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [2]:
import numpy as np
import pandas as pd
from scipy import stats

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [4]:
pokemon = pd.read_csv('/Users/alessandra/Documents/Ironhack/Curso/Data/pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [5]:
pokemon.head(10)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9  7                   Squirtle  Water     NaN    314  44      48       65   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False  
6      109       85    100           1      False  
7      130       85    100           1      False  
8      159      115    100           1      False  
9       50       64     43           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [12]:
pokemon['Legendary'].value_counts()

False    735
True      65
Name: Legendary, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [26]:
pokemon2 = pokemon[(pokemon['Legendary'] == True)].describe()[['Total']].loc[['mean', 'std'], :]

In [27]:
pokemon2.rename(columns={'Total': 'Legendary'})

Legendary
mean  637.384615
std    60.937389

In [28]:
pokemon3 = pokemon[(pokemon['Legendary'] == False)].describe()[['Total']].loc[['mean', 'std'], :]

In [29]:
pokemon3.rename(columns={'Total': 'Non-Legendary'})

Non-Legendary
mean     417.213605
std      106.760417

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [30]:
legendary = pokemon[(pokemon['Legendary'] == True)]['Total']
non_legendary = pokemon[(pokemon['Legendary'] == False)]['Total']

stats.ttest_ind(legendary, non_legendary)

Ttest_indResult(statistic=16.386116965872425, pvalue=3.0952457469652825e-52)

What do you conclude from this test? Write your conclusions below.

**Test t: when you are trying to find evidence of a significant difference between population means!**

**The t-value measures the size of the difference in relation to the variation in your sample data. The greater the magnitude of T, the greater the evidence that there is a significant difference (rejecting H0). The closer T is to 0, the greater the probability that there will be no significant difference.
The P-value is the probability of obtaining test results at least as extreme as the results actually observed during the test, assuming that the null hypothesis is correct.**

**H0: mean_legendary = mean_non_legendary / H1: mean_legendary ≠ mean_non_legendary**

**Answer: As the p-value<0.05, we could reject H0, that is, the averages of legendary and non-legendary are not equal.**
    


How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [31]:
pokemon['Type 1'].value_counts()

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ghost        32
Ground       32
Dark         31
Poison       28
Fighting     27
Steel        27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [33]:
pokemon4 = pokemon[(pokemon['Type 1'] == 'Water')].describe()[['Total']].loc[['mean', 'std'], :]
pokemon4.rename(columns={'Total': 'Water'})

Water
mean  430.455357
std   113.188266

In [35]:
pokemon5 = pokemon[(pokemon['Type 1'] != 'Water')].describe()[['Total']].loc[['mean', 'std'], :]
pokemon5.rename(columns={'Total': 'Non-Water'})

Non-Water
mean  435.859012
std   121.091682

Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [36]:
water = pokemon[(pokemon['Type 1'] == 'Water')]['Total']
non_water = pokemon[(pokemon['Type 1'] != 'Water')]['Total']

stats.ttest_ind(water, non_water)

Ttest_indResult(statistic=-0.4418547448849676, pvalue=0.6587140317488793)

Write your conclusion below.

**H0: mean_water = mean_non_water / H1: mean_water ≠ mean_non_water**

**Answer: As the p-value>0.05, we couldn't reject H0, that is, the averages of water and non-water do not seem to have a significant difference.**

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [37]:
stats.ttest_rel(pokemon['Attack'], pokemon['Defense'])

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

Describe the results of the test in the cell below.

**H0: mean_attack = mean_defense / H1: attack ≠ mean_defense**

**Answer: As the p-value<0.05, we could reject H0, that is, the averages of attack and defense are not equal.**

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [38]:
stats.ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])

Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)

Describe the results of the test in the cell below.

**H0: mean_spattack = mean_spdefense / H1: mean_spattack ≠ mean_spdefense**

**Answer: As the p-value>0.05, we couldn't reject H0, that is, the averages of special attack and special defense do not seem to have a significant difference.**

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [39]:
stats.ttest_1samp(pokemon['Sp. Atk'] - pokemon['Sp. Def'], 0)

Ttest_1sampResult(statistic=0.853986188453353, pvalue=0.3933685997548122)

**Answer: The results are the same!**

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [40]:
contigency_table = pd.crosstab(pokemon['Legendary'], pokemon['Type 1'] == 'Water')
contigency_table

Type 1     False  True 
Legendary              
False        627    108
True          61      4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [41]:
stats.chi2_contingency(contigency_table)

(2.9429200762850503, 0.08625467249550949, 1, array([[632.1, 102.9],
        [ 55.9,   9.1]]))

Based on a 95% confidence, should we reject the null hypothesis?

**Chi-Square Test: We should use the Chi-Square Test when we want to compare independent categorical variables and verify if they are homogeneous. It is used in scenarios where sample sizes are larger.**

**Answer: As the p-value>0.05 (0.086), we couldn't reject H0, that is, the averages do not seem to have a significant difference.**